esmvalcore/_data_finder.py

```diff
@@ -108,23 +187,89 @@ def get_start_end_year(filename):
     return int(start_year), int(end_year)
 
 
-def select_files(filenames, start_year, end_year):
-    """Select files containing data between start_year and end_year.
-
-    This works for filenames matching *_YYYY*-YYYY*.* or *_YYYY*.*
+def _parse_period(timerange):
+    start_date = None
+    end_date = None
+    time_format = None
+    datetime_format = (
+        isodate.DATE_BAS_COMPLETE + 'T' + isodate.TIME_BAS_COMPLETE)
+    if timerange.split('/')[0].startswith('P'):
+        try:
+            end_date = isodate.parse_datetime(timerange.split('/')[1])
+            time_format = datetime_format
+        except isodate.ISO8601Error:
+            end_date = isodate.parse_date(timerange.split('/')[1])
+            time_format = isodate.DATE_BAS_COMPLETE
+        delta = isodate.parse_duration(timerange.split('/')[0])
+        start_date = end_date - delta
+    elif timerange.split('/')[1].startswith('P'):
+        try:
+            start_date = isodate.parse_datetime(timerange.split('/')[0])
+            time_format = datetime_format
+        except isodate.ISO8601Error:
+            start_date = isodate.parse_date(timerange.split('/')[0])
+            time_format = isodate.DATE_BAS_COMPLETE
+        delta = isodate.parse_duration(timerange.split('/')[1])
+        end_date = start_date + delta
+
+    if time_format == datetime_format:
+        start_date = isodate.datetime_isoformat(
+            start_date, format=datetime_format)
+        end_date = isodate.datetime_isoformat(
+            end_date, format=datetime_format)
+    elif time_format == isodate.DATE_BAS_COMPLETE:
+        start_date = isodate.date_isoformat(start_date, format=time_format)
+        end_date = isodate.date_isoformat(end_date, format=time_format)
+
+    return start_date, end_date
+
+
+def _compare_dates(date, file_date):
```

Comment:
The name of this function is confusing, because it doesn't actually compare the dates, but rather truncates the strings representing them so they are the same length, if I understand the code correctly